cmd/snap/complete.go

```diff
@@ -28,9 +31,39 @@ func (s installedSnapName) Complete(match string) []flags.Completion {
 	return ret
 }
 
+func completeFromSortedFile(fn, match string) ([]flags.Completion, error) {
+	fd, err := os.Open(fn)
+	if err != nil {
+		return nil, err
+	}
+	defer fd.Close()
+
+	var ret []flags.Completion
+
+	// TODO: look into implementing binary search
+	//       e.g. https://github.com/pts/pts-line-bisect/
+	scanner := bufio.NewScanner(fd)
+	for scanner.Scan() {
```

Comment:
Probably good enough for ~2k package names, but we should look into a more efficient format for commands or once we get too many items.

Reply:
it's good enough well beyond that:
```
~$ wc -l /usr/share/dict/british-english-insane 
650656 /usr/share/dict/british-english-insane
~$ time /tmp/completeFromSortedFile /usr/share/dict/british-english-insane abo | wc -l
<nil>
190

real	0m0.025s
user	0m0.016s
sys	0m0.008s
```

the limit is somewhere in the 1M range:
```
~$ wc -l /var/tmp/out
10000001 /var/tmp/out
~$ time /tmp/completeFromSortedFile /var/tmp/out abo | wc -l
<nil>
2389706

real	0m5.282s
user	0m3.944s
sys	0m4.836s
```

but then we can bisect:

```
/tmp/pts-line-bisect$ time ./pts_lbsearch -pc /usr/share/dict/british-english-insane abo | wc -l
190

real	0m0.003s
user	0m0.000s
sys	0m0.000s
/tmp/pts-line-bisect$ time ./pts_lbsearch -pc /var/tmp/out abo | wc -l
2389706

real	0m0.041s
user	0m0.032s
sys	0m0.036s
```

given that the android and ios appstores both have less than 3M apps, I think having something that works well into that range of values is good enough, period.

Reply:
average times on a bbb with the 500k wordlist are ~.5s, which is a little too much but still faster than the remote completion we have today. With the 10M wordlist, it's ~2 minutes. Both are ~.05s with bisection, but can grow to be ~1m if there are too many matches. So when we reach that many entries we probably want to not print all matches!

Reply:
Wow, nice numbers. Thanks a lot for the tests!

Do you have an idea of how many commands we're supposed to expect for the command-not-found use?

Reply:
not more than an order of magnitude more